doc/rewriting-services.md

```diff
@@ -0,0 +1,204 @@
+
+```js
+const BaseJsonService = require('../base-json')
+
+class ExampleDownloads extends BaseJsonService {
+  static get route() {
+    return {
+       base: 'example/d',
+       pattern: ':param1/:param2’:',
+     }
+   }
+
+   static defaultBadgeData() {
+     return { label: 'downloads' } // or whatever
+   }
+
+   async handle({ param1, param2 }) {
+     return { message: 'hello' }
+   }
+}
+```
+
+4. We don’t have really good tools for debugging matches, so the best you can do
+is run a subset of your tests. To run a single service test, add `.only()`
+somewhere in the chain, and run `npm run test:services:trace — —only=example`.
+
+```js
+t.create('build status')
+  .get('/pip.json')
+  .only() // Prevent this ServiceTester from running its other tests.
+  .expectJSONTypes(
+    Joi.object().keys({
+      name: 'docs',
+      value: Joi.alternatives().try(isBuildStatus, Joi.equal('unknown')),
+    })
+  )
+```
+
+5. Presumably the test will fail, though by examining the copious output, you
+can confirm the route was matched and the named parameters mapped successfully.
+Since you'll have just run the tests on the old code (right?) you'll know you
+haven't inadvertently changed the route (an easy mistake to make).
+
+6. If the legacy service had a base URL and you've changed it, you’ll need to
+update the tests _and_ the examples. Take care to do that.
+
+## Implement `render()` and `handle()`
+
+1. Once the route is working, fill out `render()` and `handle()`. If there’s a
+single service, you can implement fetch as a method or a function at the top of
+the file. If there are more than one service which share fetching code, you can
+put the fetch function in `example-common.js` like this one for github:
+
+<details>
+
+```js
+const Joi = require('joi')
+const { errorMessagesFor } = require('./github-helpers')
+
+const issueSchema = Joi.object({
+  head: Joi.object({
+    sha: Joi.string().required(),
+  }).required(),
+}).required()
+
+async function fetchIssue(serviceInstance, { user, repo, number }) {
+  return serviceInstance._requestJson({
+    schema: issueSchema,
+    url: `/repos/${user}/${repo}/pulls/${number}`,
+    errorMessages: errorMessagesFor('pull request or repo not found'),
+  })
+}
+
+module.exports = {
+  fetchIssue,
+}
+```
+
+</details>
+
+or create an abstract superclass like **PypiBase**:
+
+<details>
+
+```js
+const Joi = require('joi')
+const BaseJsonService = require('../base-json')
+
+const schema = Joi.object({
+  info: Joi.object({
+    version: Joi.string().required(),
+    // https://github.com/badges/shields/issues/2022
+    license: Joi.string().allow(''),
+    classifiers: Joi.array()
+      .items(Joi.string())
+      .required(),
+  }).required(),
+  releases: Joi.object()
+    .pattern(
+      Joi.string(),
+      Joi.array()
+        .items(
+          Joi.object({
+            packagetype: Joi.string().required(),
+          })
+        )
+        .required()
```

Comment:
This schema object is incomplete. Also the majority of this code snippet is the schema definition but it isn't really the main thing we're trying to demonstrate. Can we pick an example that uses a less verbose schema so the example is clearer.

Reply:
Is there another good one to show? I dug around a bit before settling on this one. I'll snip the schema for now, though feel free to replace it with a better example (or suggest one, and I'll update).